pkg/serviceaccount/externaljwt/metrics/metrics.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "externaljwt"
+)
+
+var (
+	lastKeyFetchTimeStamp = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "supported_keys_last_fetch_time",
+			Help:           "Timestamp of the last successful key sync",
+			StabilityLevel: metrics.ALPHA,
+		},
+		nil,
+	)
+
+	dataTimeStamp = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "supported_keys_data_timestamp",
+			Help:           "Unix Timestamp in seconds associated with snapshot of supported keys as returned by external signer",
+			StabilityLevel: metrics.ALPHA,
+		},
+		nil,
+	)
+
+	totalKeyFetch = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "fetch_keys_request_total",
+			Help:           "Total attempts at syncing supported JWKs",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"code"},
+	)
+
+	tokenGenReqTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "sign_request_total",
+			Help:           "Total attempts at signing JWT",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"code"},
+	)
+
+	outboundRequest = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "outbound_request",
+			Help:           "Request duration and time for calls to external-jwt-signer",
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
```

Comment:
I don't have a good sense for what buckets make sense for this... @enj, any feedback from what we selected for KMS?

```
			// Use custom buckets to avoid the default buckets which are too small for KMS operations.
			// Start 0.1ms with the last bucket being [~52s, +Inf)
			Buckets: metrics.ExponentialBuckets(0.0001, 2, 20),
```

if we left these unset, we'd get defaults:
```
DefBuckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}
```

those actually look plausibly close to me... I'd maybe just tack on a 1ms at the front and 1-2 longer ones at the end, and explicitly enumerate, like:

```go
[]float64{.001, .005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 30, 60}
```

(if we don't explicitly enumerate, it's helpful to document the min/max bucket size in a comment... doing mental math to know what these are is hard)

Reply:
added `[]float64{.001, .005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 30, 60}`